protocols/autonat/src/behaviour.rs

```diff
@@ -512,3 +521,97 @@ trait HandleInnerEvent {
         event: RequestResponseEvent<DialRequest, DialResponse>,
     ) -> (VecDeque<Event>, Option<Action>);
 }
+
+trait GlobalIp {
+    fn is_global_ip(&self) -> bool;
+}
+
+impl GlobalIp for Multiaddr {
+    fn is_global_ip(&self) -> bool {
+        match self.iter().next() {
+            Some(Protocol::Ip4(a)) => a.is_global_ip(),
+            Some(Protocol::Ip6(a)) => a.is_global_ip(),
+            _ => false,
+        }
+    }
+}
+
+impl GlobalIp for std::net::Ipv4Addr {
+    // NOTE: The below logic is copied from `std::net::Ipv4Addr::is_global`, which is at the time of
+    // writing behind the unstable `ip` feature.
+    // See https://github.com/rust-lang/rust/issues/27709 for more info.
+    //
+    // TODO: Consider removing check check with the unstable methods `is_shared`, `is_reserved` and `is_benchmarking`? In case
+    // of the former the peer is still eligible for a dial-back, the latter cases should never occur in practice.
+    //
+    // TODO: Consider to only check for loopback, private ip and link-local. Can any other case ever happen in practice if we solely
+    // use this check for observed addresses?
```

Comment:
I don't think it is worth deviating from the `IPv4Addr::is_global` function for the sake of not copying the code below. I don't expect it to change any time soon. Am I missing something?

Reply:
It would be for the sake of avoiding unstable methods. `is_shared`, `is_reserved` and `is_benchmarking` are the only non-stable methods calls within the `is_global` implementation. In particular for `is_reserved` there is a warning:
> As IANA assigns new addresses, this logic will be updated. This may result in non-reserved addresses being treated as reserved in code that relies on an outdated version of this method.

If those are not relevant for our case anyway, we would avoid the risk of being outdated if the std implementation changes.

Reply:
Off the top of my head I would favor consistency with `std` over avoiding the risk of being outdated. I.e. this code being an exact copy of `std` makes it easier to maintain.